sklearn/tree/_tree.pyx

```diff
@@ -1230,3 +1228,399 @@ cdef class Tree:
             if not (0.999 < total_weight < 1.001):
                 raise ValueError("Total weight should be 1.0 but was %.9f" %
                                  total_weight)
+
+
+# =============================================================================
+# Build Pruned Tree
+# =============================================================================
+
+
+cdef class _CCPPruneController:
```

Comment:
curious design question, why not pass a function with a certain signature instead of inheriting a class?

Specially since you have two classes inheriting this one, and each override only one of those functions, it seems natural to me to just pass a function. WDYT?

Reply:
Hmmm are you suggesting passing two functions since `stop_pruning` and `save_metrics` need to be called at different times?

The controller almost feels like callbacks. It contains state and needs to be called at different places during pruning.